### PR TITLE
feat(fortuna-v2): implement a keeper service for entropy v2

### PR DESCRIPTION
**Pull request by 0xfirefist**

### CODE REVIEW

File: fortuna/src/keeper.rs
Comment (m30m) [inline fortuna/src/keeper.rs:263]: why not websocket?

File: fortuna/src/command/run.rs
Comment (m30m) [inline fortuna/src/command/run.rs:119]: await on changed

File: fortuna/src/command/run.rs
Comment (m30m) [inline fortuna/src/command/run.rs:131]: after looking at all the logic, it seems that we don't essentially need graceful shutdown on the keeper side. I think we can remove the exit checks in this path.

File: fortuna/src/config/run.rs
Comment (m30m) [inline fortuna/src/config/run.rs:37]: maybe `keeper-private-key`?
Also mention that this key should not have any specific privilege but just used for submitting transactions.

File: fortuna/src/keeper.rs
Comment (0xfirefist) [reply fortuna/src/keeper.rs:266]: we don't have wss:://{links}

File: fortuna/src/keeper.rs
Comment (0xfirefist) [reply fortuna/src/keeper.rs]: yeah agree. I added this when I didn't thought of simulating a request. Removed

File: fortuna/src/keeper.rs
Comment (m30m) [inline fortuna/src/keeper.rs:64]: we should use nonce manager to send these transactions instead of manually getting the next nonce

File: fortuna/src/keeper.rs
Comment (m30m) [inline fortuna/src/keeper.rs:112]: use `saturating_sub`

File: fortuna/src/keeper.rs
Comment (m30m) [inline fortuna/src/keeper.rs:121]: why do we need this?

File: fortuna/src/keeper.rs
Comment (m30m) [inline fortuna/src/keeper.rs:119]: shouldn't this be `from_block <= last_block`?

File: fortuna/src/keeper.rs
Comment (m30m) [inline fortuna/src/keeper.rs:158]: add `to block` here as well

File: fortuna/src/keeper.rs
Comment (m30m) [inline fortuna/src/keeper.rs:251]: you can merge these two

File: fortuna/src/keeper.rs
Comment (0xfirefist) [reply fortuna/src/keeper.rs:119]: good catch. thanks!

File: fortuna/src/command/run.rs
Comment (jayantk) [inline fortuna/src/command/run.rs:110]: nitpick: I think you could probably set up the parsing logic that creates the hashmap such that the values of the hashmap contain enough information to spawn the keeper threads. That would make this method less error prone, because callers couldn't accidentally pass a mismatched Config and HashMap

File: fortuna/src/config/run.rs
Comment (jayantk) [inline fortuna/src/config/run.rs:37]: "specific priviledges" is a bit vague. Suggest saying "This key should not be a registered provider"

File: fortuna/src/keeper.rs
Comment (jayantk) [inline fortuna/src/keeper.rs:50]: shouldn't this be called chain_state

File: fortuna/src/keeper.rs
Comment (jayantk) [inline fortuna/src/keeper.rs:69]: can you document this method please? this is a good spot to explain what the different keeper threads are and what they do

File: fortuna/src/keeper.rs
Comment (jayantk) [inline fortuna/src/keeper.rs:42]: in the future, this may need to be configurable. I think some chains like arbitrum have a blocktime that's < 1 second (because every tx is a block)

File: fortuna/src/command/run.rs
Comment (jayantk) [reply fortuna/src/command/run.rs:110]: After reading through the code below a bit, one way you could do this is by creating a trait like `EntropyWriter` that's analogous to `EntropyReader` and then adding that to `BlockchainState`. This trait would also help you write tests for the new logic (that's why I created EntropyReader in the first place, actually).

I wouldn't worry about this now though. Just something to think about when writing code in the future.

File: fortuna/src/keeper.rs
Comment (jayantk) [inline fortuna/src/keeper.rs:190]: does this estimate need to be padded a bit to ensure the TX lands? (I really don't know the answer to this)

File: fortuna/src/keeper.rs
Comment (jayantk) [reply fortuna/src/keeper.rs:69]: some inline comments explaining the logic would help too, e.g., watch_blocks returns a stream of new blocks that get processed by process_new_blocks . In this particular case, it's worth explaining *why* the code works like this. Why subscribe to new blocks vs the events directly? There must be some thinking behind that. Write down the reasoning in the code for future reference.

Comment (jayantk) [reply fortuna/src/keeper.rs:69]: actually generally think all of the functions in this file could use a nice doc comment on top. The current comments are either missing or aren't descriptive enough imo.

**Review by jayantk — APPROVED** [review]
Nice, I think this looks pretty good on the whole. I think you probably could refactor this in the future to enable unit testing, but no need to let that block this PR. I left a couple requests for code comments, so please look at those. Feel free to merge without another review after addressing those comments.

File: fortuna/Cargo.toml
Comment (jayantk) [inline fortuna/Cargo.toml:3]: 4.0.0

**Pull request merged**
